### PR TITLE
feat: offline improvements

### PR DESCRIPTION
**Pull request by sebastianquek**
* Overall improvements to handle the offline state

### Issues
1. The types don't work properly for the fixtures since they don't have a `version` attribute. For cases where any of these fixtures are used, I've casted them to `any` for now since it's likely we don't want the `version` attribute to be required.
2. `DocumentListScreenContainer`: `Property 'name' does not exist on type 'OpenAttestationDocument'.`
3.  ~~`fetchDocument`: `Could not find a declaration file for module '@govtechsg/oa-encryption'.`~~
4. Need to figure out a way to "cancel" previous verify function calls in `useDocumentVerifier`
  a. So in the case when `verify` is called multiple times, we only set state based on the last call
  b. Can be handled typically in `useEffect` where the return function can cancel the setting of state for that closure, but we're using `useCallback` here, so not sure.
  c. Right now this is all circumvented by disabling `exhaustive-deps`, so that `verify` is only called when `isConnected` changes. Ideally we don't need this.

### Changes
```
chore: upgrade oa-verify and open-attestation
feat: make use of new oa-verify API
chore: remove unnecessary IdentityVerifier
chore: use WrappedDocument and v2 oa document types
feat: expose a verify function for useDocumentVerifier so that we can call it imperatively
feat: call verify when network conditions change and is connected
f̶e̶a̶t̶:̶ ̶a̶d̶d̶ ̶E̶r̶r̶o̶r̶ ̶C̶h̶e̶c̶k̶S̶t̶a̶t̶u̶s̶
feat: show offline messages within the ValidityBanner when user is offline
feat: update isVerified state only when checkStatus is valid or invalid
feat: getIssuerName can return undefined
```

### CODE REVIEW

**Comment by john-dot-oa** [comment]
## Application
Expo QR
Published to https://exp.host/@dlt/identitywallet?release-channel=60

## Storybook
Expo QR
Published to https://exp.host/@dlt/identitywallet?release-channel=storybook-60

**Comment by john-dot-oa** [comment]
## Application
Expo QR
Published to https://exp.host/@dlt/identitywallet?release-channel=60

## Storybook
Expo QR
Published to https://exp.host/@dlt/identitywallet?release-channel=storybook-60

File: src/components/QrScanner/ValidityCheckScreenContainer.test.tsx [hunk 1]
```diff
@@ -42,7 +48,10 @@ describe("ValidityCheckScreenContainer", () => {
 
   it("should navigate when document is valid", () => {
     expect.assertions(1);
-    mockUseVerifier.mockReturnValue({ overallValidity: CheckStatus.VALID });
+    mockUseVerifier.mockReturnValue({
+      statuses: { overallValidity: CheckStatus.VALID },
+      verify: jest.fn()
+    });
 
     render(<ValidityCheckScreenContainer navigation={mockNavigation} />);
```
Comment (yehjxraymond) [inline src/components/QrScanner/ValidityCheckScreenContainer.test.tsx]: should we now include a test that checks that `verify` is called when the component mounts?

File: src/components/QrScanner/ValidityCheckScreenContainer.tsx [hunk 2]
```diff
@@ -11,27 +10,32 @@ import { replaceRouteFn } from "../../common/navigation";
...
 }) => {
-  const document: SignedDocument = navigation.getParam("document");
+  const document: OAWrappedDocument = navigation.getParam("document");
   const isSavable: boolean = navigation.getParam("savable");
-  const verificationStatuses = useDocumentVerifier(document as SignedDocument);
+  const { statuses, verify } = useDocumentVerifier();
+
+  useEffect(() => {
+    verify(document);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (yehjxraymond) [inline src/components/QrScanner/ValidityCheckScreenContainer.tsx]: actually is this required? if we put `verify` and `document` inside will it result in a loop?

File: src/services/DocumentVerifier.ts [hunk 3]
```diff
@@ -1,52 +1,36 @@
...
+} from "@govtechsg/oa-verify";
+import { OAWrappedDocument } from "../types";
 
-// Ensures TS infers the type of the array as a tuple instead
-// https://github.com/Microsoft/TypeScript/pull/24897
-function tuple<T extends any[]>(...data: T): T {
-  return data;
-}
+const verifiers = [
+  ...openAttestationVerifiers.slice(0, 2),
```
Comment (yehjxraymond) [inline src/services/DocumentVerifier.ts]: Will be nice to explain why 0,2 and 3,5 in this code.

File: src/services/DocumentVerifier.ts [hunk 4]
```diff
@@ -1,52 +1,36 @@
...
-    .catch(() => false);
-
-  return tuple(
-    verifyHash,
-    verifyIssued,
-    verifyRevoked,
-    verifyIdentity,
-    overallValidityCheck
-  );
+  const statuses = overallResult.map(f => f.status);
```
Comment (yehjxraymond) [inline src/services/DocumentVerifier.ts]: oa-verify export this: https://github.com/Open-Attestation/oa-verify#isvalid

Comment (sebastianquek) [reply src/services/DocumentVerifier.ts]: Hmm yea, thing is I wanted none of the tests to skip (ISSUER_IDENTITY was being skipped for one of our tests files), so in the case that a test skips, then it should return `ERROR`.

Comment (sebastianquek) [reply src/services/DocumentVerifier.ts]: Will use `isValid` for now and handle error states in future perhaps within `oa-verify`

File: src/components/QrScanner/ValidityCheckScreenContainer.tsx (hunk 2 above)
Comment (sebastianquek) [reply src/components/QrScanner/ValidityCheckScreenContainer.tsx]: Turns out this isn't necessary

**Comment by john-dot-oa** [comment]
## Application
Expo QR
Published to https://exp.host/@dlt/identitywallet?release-channel=60

## Storybook
Expo QR
Published to https://exp.host/@dlt/identitywallet?release-channel=storybook-60

**Comment by john-dot-oa** [comment]
## Application
Expo QR
Published to https://exp.host/@dlt/identitywallet?release-channel=60

## Storybook
Expo QR
Published to https://exp.host/@dlt/identitywallet?release-channel=storybook-60

File: package.json [hunk 5]
```diff
@@ -15,7 +15,7 @@
   "dependencies": {
     "@govtechsg/dnsprove": "^2.0.5",
     "@govtechsg/oa-encryption": "^1.3.0",
-    "@govtechsg/oa-verify": "^2.1.0",
+    "@govtechsg/oa-verify": "^3.0.3",
```
Comment (Nebulis) [inline package.json]: there is nothing to prevent merging while types are wrong in the project. I suggest to add a step to check types and run it in the CI

File: src/components/DocumentList/DocumentListScreenContainer.tsx [hunk 6]
```diff
@@ -5,7 +5,7 @@ import { useDbContext } from "../../context/db";
 import { replaceRouteFn } from "../../common/navigation";
 import { DocumentListScreen } from "./DocumentListScreen";
 
-const getIssuerName = (document: DocumentObject): string => {
+const getIssuerName = (document: DocumentObject): string | undefined => {
```
Comment (Nebulis) [inline src/components/DocumentList/DocumentListScreenContainer.tsx]: there is a helper to do that in oa FYI

File: src/components/DocumentRenderer/LocalDocumentRendererContainer.tsx [hunk 7]
```diff
@@ -29,8 +29,13 @@ export const LocalDocumentRendererContainer: FunctionComponent<NavigationProps>
     const updateFunction = (
       oldData: DocumentProperties
     ): DocumentProperties => {
-      oldData.isVerified = checkStatus === CheckStatus.VALID;
-      oldData.verified = Date.now();
+      if (checkStatus === CheckStatus.VALID) {
+        oldData.isVerified = true;
+        oldData.verified = Date.now();
+      } else if (checkStatus === CheckStatus.INVALID) {
```
Comment (Nebulis) [inline src/components/DocumentRenderer/LocalDocumentRendererContainer.tsx]: Do you want to also handle ERROR here for the moment, or just ignoring ?

Comment (Nebulis) [reply src/components/DocumentRenderer/LocalDocumentRendererContainer.tsx]: looks done by handleVerificationFragment but might be hard to follow :D

File: package.json (hunk 5 above)
Comment (Nebulis) [reply package.json]: image

All this files have TS errors

File: src/components/DocumentRenderer/LocalDocumentRendererContainer.tsx (hunk 7 above)
Comment (Nebulis) [reply src/components/DocumentRenderer/LocalDocumentRendererContainer.tsx]: ok no it's overall status :see_no_evil:

File: src/common/hooks/useDocumentVerifier/index.tsx [hunk 8]
```diff
@@ -1,8 +1,16 @@
...
+  useState,
+  Dispatch,
+  SetStateAction,
+  useCallback,
+  useRef
+} from "react";
 import { CheckStatus } from "../../../components/Validity";
 import { checkValidity } from "../../../services/DocumentVerifier";
 import { useConfigContext } from "../../../context/config";
+import { VerificationFragmentStatus } from "@govtechsg/oa-verify";
```
Comment (benjaminlim00) [inline src/common/hooks/useDocumentVerifier/index.tsx]: no such import from the version of oa-verify that is being used

File: src/components/DocumentList/DocumentListScreenContainer.tsx (hunk 6 above)
Comment (sebastianquek) [reply src/components/DocumentList/DocumentListScreenContainer.tsx]: Ok, we can use that in a separate PR

File: .circleci/config.yml [hunk 9]
```diff
@@ -8,6 +8,9 @@ jobs:
       - run:
           name: install
           command: npm install
+      - run:
+          name: tsc
+          command: "./node_modules/typescript/bin/tsc"
```
Comment (Nebulis) [inline .circleci/config.yml]:
```suggestion
          command: npx --no-install tsc --noEmit
```

**Comment by john-dot-oa** [comment]
## Application
Expo QR
Published to https://exp.host/@dlt/identitywallet?release-channel=60

## Storybook
Expo QR
Published to https://exp.host/@dlt/identitywallet?release-channel=storybook-60
